From d3ff3c0af397f9318ae8b041842b3e3b9257271b Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Thu, 19 Sep 2024 12:53:33 +0800
Subject: [PATCH 213/265] mtk: mt76: mt7996: check addba api return value

Clear amptu_stat bit if ieee80211_start_tx_ba_session return fail.
Without this patch, driver does not know ba is not established and would
not send addba anymore.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt7996/mac.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/mt7996/mac.c b/mt7996/mac.c
index 46915769..7effbdd5 100644
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -1157,8 +1157,10 @@ mt7996_tx_check_aggr(struct ieee80211_sta *sta, struct sk_buff *skb,
 	if (!mlink)
 		return;
 
-	if (!test_and_set_bit(tid, &mlink->wcid.ampdu_state))
-		ieee80211_start_tx_ba_session(sta, tid, 0);
+	if (!test_and_set_bit(tid, &mlink->wcid.ampdu_state)) {
+		if (ieee80211_start_tx_ba_session(sta, tid, 0))
+			clear_bit(tid, &mlink->wcid.ampdu_state);
+	}
 }
 
 static void
-- 
2.45.2

